c7n/resources/elasticsearch.py

```diff
@@ -622,3 +624,114 @@ def process(self, resources):
+        "Effect": "Allow",
+        "Principal": {"Service": ["es.amazonaws.com"]},
+        "Action": ["logs:PutLogEvents", "logs:CreateLogStream"],
+        "Resource": None}
+
+    permissions = (
+        'es:UpdateElasticsearchDomainConfig',
+        'es:ListDomainNames',
+        'logs:DescribeLogGroups',
+        'logs:CreateLogGroup',
+        'logs:PutResourcePolicy')
+
+    def get_loggroup_arn(self, domain, log_prefix=None):
+        if log_prefix:
+            log_group_name = "%s/%s/audit-logs" % (log_prefix, domain)
+        else:
+            log_group_name = "/aws/OpenSearchService/domains/%s/audit-logs" % (domain)
+        log_group_arn = "arn:aws:logs:{}:{}:log-group:{}:*".format(
+            self.manager.region, self.manager.account_id, log_group_name)
+        return log_group_arn
+
+    def merge_dict(self, d1, d2):
+        merged_dict = defaultdict(list)
+
+        for d in (d1, d2):
+            for key, value in d.items():
+                if key == 'Resource':
+                    if isinstance(value, list):
+                        merged_dict[key].extend(value)
+                    else:
+                        merged_dict[key].append(value)
+                else:
+                    merged_dict[key] = value
+
+        return dict(merged_dict)
+
+    def set_permissions(self, log_group_arn):
+        statement = dict(self.statement)
+        statement['Resource'] = [log_group_arn]
+        client = local_session(
+            self.manager.session_factory).client('logs')
+
+        try:
+            client.create_log_group(logGroupName=log_group_arn.split(":")[-2])
+        except client.exceptions.ResourceAlreadyExistsException:
+            self.log.warning("Log group already exists ...")
+
+        for policy in client.describe_resource_policies().get('resourcePolicies'):
+            if policy['policyName'] == "OpenSearchCloudwatchLogPermissions":
+                policy_doc = json.loads(policy['policyDocument'])
+                if log_group_arn not in policy_doc['Statement'][0]['Resource']:
+                    merged_statement = self.merge_dict(statement, policy_doc['Statement'][0])
+                    statement = merged_statement
+                continue
+
+        response = client.put_resource_policy(
+            policyName='OpenSearchCloudwatchLogPermissions',
+            policyDocument=json.dumps(
+                {"Version": "2012-10-17", "Statement": statement}))
+
+        return response
+
+    def process(self, resources):
+        client = local_session(
+            self.manager.session_factory).client('es')
+        state = self.data.get('state')
+        log_prefix = self.data.get('loggroup_prefix')
+        log_group_arns = {
+            r['DomainName']: self.get_loggroup_arn(r["DomainName"], log_prefix) for r in resources}
+
+        for r in resources:
+            if state:
+                self.set_permissions(log_group_arns[r["DomainName"]])
+                time.sleep(15)
```

Comment:
setting the delay as a class attribute will allow us to avoid a slow test on state, since we can patch the value in the test.

Reply:
fixed ... great suggestion.